drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,421 @@
+					       min_uv, max_uv, &idx);
+	if (ret == -EINVAL) {
+		LOG_INST_ERR(config->log, "No voltage range window could be deteceted");
+		return ret;
+	}
+
+	idx <<= config->desc->vsel_bitpos;
+
+	LOG_INST_DBG(config->log, "Setting voltage=%d", config->modes_uv);
+	LOG_INST_DBG(config->log, "[0x%x]=0x%x mask=0x%x", (uint32_t)config->desc->vsel_reg, idx,
+		     (uint32_t)config->desc->vsel_mask);
+	ret = i2c_reg_update_byte_dt(&cconfig->i2c, config->desc->vsel_reg, config->desc->vsel_mask,
+				     (uint8_t)idx);
+
+	if (0 != ret) {
+		LOG_INST_ERR(config->log, "Failed to set regulator voltage");
+	}
+
+	return ret;
+}
+
+int axp192_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->parent->config;
+	int ret;
+	uint8_t raw_reg;
+
+	ret = i2c_reg_read_byte_dt(&cconfig->i2c, config->desc->enable_reg, &raw_reg);
+
+	if (0u == (raw_reg & config->desc->enable_mask)) {
+		/* regulator disabled */
+		*volt_uv = 0u;
+	} else {
+		ret = i2c_reg_read_byte_dt(&cconfig->i2c, config->desc->vsel_reg, &raw_reg);
+		if (ret < 0) {
+			return ret;
+		}
+
+		raw_reg = (raw_reg & config->desc->vsel_mask) >> config->desc->vsel_bitpos;
+
+		ret = linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges,
+						   raw_reg, volt_uv);
+	}
+
+	return ret;
+}
+
+int axp192_get_current_limit(const struct device *dev, int32_t *curr_ua)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	*curr_ua = config->desc->max_ua;
+
+	return 0;
+}
+
+static struct regulator_driver_api api = {
+	.enable = axp192_enable,
+	.disable = axp192_disable,
+	.count_voltages = axp192_count_voltages,
+	.list_voltage = axp192_list_voltage,
+	.set_voltage = axp192_set_voltage,
+	.get_voltage = axp192_get_voltage,
+	.get_current_limit = axp192_get_current_limit,
+};
+
+static int regulator_axp192_init(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	int ret = 0;
+
+	regulator_common_data_init(dev);
+
+	LOG_INST_DBG(config->log, "voltage=%d", config->modes_uv);
+
+	if (!device_is_ready(config->parent)) {
+		LOG_INST_ERR(config->log, "Parent instance not ready!");
+		return -ENODEV;
+	}
+
+	/* configure voltage */
+	if (config->modes_uv == 0) {
+
+		/* disable regulator if voltage is 0 */
+		ret = axp192_disable(dev);
+		if (ret < 0) {
+			return ret;
+		}
+
+	} else if (config->modes_uv > 0) {
+
+		/* Make sure regulator is enabled */
+		ret = axp192_enable(dev);
+		if (ret < 0) {
+			return ret;
+		}
+
+		/* program mode voltage */
+		ret = axp192_set_voltage(dev, config->modes_uv, config->modes_uv);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return regulator_common_init(dev, false);
+}
+
+static int regulator_pxa192_common_init(const struct device *dev)
+{
+	return 0;
+}
+
+int regulator_axp192_dvs_state_set(const struct device *dev, regulator_dvs_state_t state)
```

Comment:
You can omit dvs_state_set and the parent_api definitions if they are not supported, it will be handled by the common interface:
https://github.com/zephyrproject-rtos/zephyr/blob/7f2a7ed2de585879ad12150f52b71dad1c2873d5/include/zephyr/drivers/regulator.h#L275

Reply:
Done